### PR TITLE
Implement MOPD for the `DistillationTrainer`

### PR DESCRIPTION
**Pull request by cmpatino**
# What does this PR do?

Adds multi-teacher on-policy distillation (MOPD, https://huggingface.co/papers/2606.30406) to the`DistillationTrainer` as an opt-in `teacher_models` argument. Each dataset row selects its teacher through a `teacher_id` column, and the student is distilled against that teacher's full next-token distribution with the existing chunked divergence loss.

The single-teacher `teacher_model` path is unchanged. Registering one teacher through`teacher_models` is bitwise identical to `teacher_model=` (measured on single GPU, DDP and DeepSpeed ZeRO-2).

## Design
- Teachers live on CPU. At each generation boundary, every teacher present in the batch is moved to the accelerator once, scores its rows for all microbatches, and is moved back. Only the completion hidden states are kept on CPU. Device memory during scoring is therefore bounded by the largest teacher, not by the sum of teachers.
- The loss splits each microbatch's rows by teacher and calls the existing `_chunked_divergence_loss` once per group, uploading each teacher's output head as needed. Only the heads of the teachers present in the microbatch are on the device during the loss. Row order is never changed.
- Each checkpoint carries `teacher_manifest.json` (id, source, resolved commit, dtype). Resuming with a different teacher set under the same ids raises.

## Assumptions and restrictions
- All teachers fit in host RAM at once, together with one generation batch of hidden-state
  targets (`per_device_train_batch_size × gradient_accumulation_steps × max_completion_length × teacher hidden_size × dtype bytes`).
- Every teacher shares the student's tokenizer (compared by canonical serialization, padding configuration excluded) and vocabulary size.
- One teacher per row; no aggregation across teachers and no per-teacher loss weights.
- Teachers must be unquantized and loaded without device dispatch or offload hooks; quantized and dispatched teacher backends have not been integrated or validated. `device_map=None` is allowed.
- Text-only students; VLM students keep using `teacher_model`.
- DeepSpeed ZeRO-3 is rejected: the teacher head is a plain device tensor and per-teacher loss groups would issue a rank-dependent number of collectives. ZeRO-1/2 and DDP are supported; under a DeepSpeed mixed-precision plugin, teachers are cast to the engine dtype (parameters only, as the engine does).

Feature requested in https://github.com/huggingface/trl/issues/6151

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [ ] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [x] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Overview**
> Unable to generate summary.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit ed46483c8f356ac7606de713a2e5a35989129d62. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/distillation_trainer.py [hunk 1]
```diff
@@ -406,12 +452,28 @@ def __init__(
         quantization_config: "BitsAndBytesConfig | None" = None,
         peft_config: "PeftConfig | None" = None,
         tools: list[Callable] | None = None,
+        teacher_models: "dict[str, str | PreTrainedModel] | None" = None,
```
Comment (qgallouedec) [inline trl/trainer/distillation_trainer.py]: api-wise, I feel it's strange to have `teacher_model` and `teacher_models`. Why not simply re-using `teacher_model` and support list/dict?

File: docs/source/paper_index.md [hunk 2]
```diff
@@ -1678,6 +1678,8 @@ training_args = GKDConfig(
...
 
+The synchronous [`DistillationTrainer`] implements the same fusion stage for locally loaded teachers: passing more than one entry in `teacher_models` enables MOPD, with each row's `teacher_id` column selecting which (already-trained, frozen) teacher scores it. See [Multiple teachers (MOPD)](distillation_trainer#multiple-teachers-mopd).
```
Comment (diogomsmiranda) [inline docs/source/paper_index.md]: Another comment mentions this and I agree that it might make more sense to use this as the behaviour for teacher_model (maybe renaming this?) instead of adding another argument like you have described here: https://github.com/huggingface/trl/pull/7215/changes#diff-3627d7ea9d494ee0001a18064dd53255aed938151295a4bb3af0c0e7cc23b8feR269 

You say that to enable MOPD you just have to instantiate `teacher_models` instead of `teacher_model`. But what happens when `teacher_models={}` or when it's 1 model, does it default back to 1 teacher kd?

File: trl/trainer/distillation_trainer.py [hunk 3]
```diff
@@ -406,12 +417,27 @@ def __init__(
...
     ):
         if args is None:
             model_name = model if isinstance(model, str) else get_config_model_id(model.config)
             model_name = model_name.split("/")[-1]
             args = DistillationConfig(f"{model_name}-Distillation")
 
+        # Multi-teacher mode is opted into by `teacher_models` and is exclusive with the singular teacher entry
+        # points, so a run can never carry two teacher sources with different lifecycles. Checked before any loading.
+        if teacher_models is not None and (teacher_model is not None or args.teacher_model_name_or_path is not None):
```
Comment (diogomsmiranda) [inline trl/trainer/distillation_trainer.py]: I think you might be missing a check somewhere here to see if `teacher_models` is not empty. Else you are letting it fail on teacher_id verification or if the dataset doesn't have one then there's no fail guard.

File: trl/trainer/distillation_trainer.py [hunk 4]
```diff
@@ -1853,6 +2118,91 @@ def _compute_loss(self, unwrapped_student, inputs, num_items_in_batch):
...
+            loss = student_hidden_states.new_zeros((), dtype=torch.float32)
+            entropy_sum = student_hidden_states.new_zeros((), dtype=torch.float32)
+            n_valid = student_hidden_states.new_zeros((), dtype=torch.int64)
+            teacher_stats = student_hidden_states.new_zeros((2, len(self.teacher_models)), dtype=torch.float32)
+            # The group losses are summed, so they must share one denominator. `num_items_in_batch` is already one;
+            # where it is `None` (evaluation) the loss would otherwise normalize each group by its own valid-token
+            # count, and the sum of the group means would scale with the number of teachers in the microbatch. The
+            # count matches the one the loss computes from the same mask, so a single group reduces to the legacy
+            # `mean` exactly.
+            denominator = loss_mask.sum().clamp(min=1) if num_items_in_batch is None else num_items_in_batch
```
Comment (diogomsmiranda) [inline trl/trainer/distillation_trainer.py:2199]: Hey, I revisited the paper that you are using to base this implementation I had a question about the normalisation relative to the paper's Eq. 1. The paper first normalizes each trajectory by its own valid length and then averages trajectories however here you normalize the total divergence by the total valid-token count instead.

So trajectory 'i' receives weight proportional to number of valid tokens, rather than uniform weight 1/B with B being the number of trajectories in the batch.  Routed domains might have systematically different trajectory lengths, bigger trajectories will have higher weights. Although this is equivalent when the size of batches per device equals 1 when >1 it might produce unwanted multi-teacher specific behaviour.

Is this intentional to be coherent with the current `DistillationTrainer`?

Comment (diogomsmiranda) [reply trl/trainer/distillation_trainer.py:2199]: For this I believe you would need to track the number of trajectories per batch which you might be able to do in similar faction to what you do with `num_items_in_batch`?

File: trl/trainer/distillation_trainer.py (hunk 1 above)
Comment (cmpatino) [reply trl/trainer/distillation_trainer.py]: I agree, so I included the change in `34874dcc`. 

`teacher_model` now takes either a single teacher (unchanged behavior) or a dict[str, str | PreTrainedModel] that routes rows by their teacher_id column. I kept it to dicts and not list/dict, since routing needs the ids.

I also updated the docstring to emphasize that the type also controls how we handle the teacher. A single teacher keeps the teacher model in the accelerator, while a mapping keeps the teachers on the CPU and moves them one at a time.

The catch is that we'd also move the teachers to a CPU even if the user specifies a mapping with one model (see the example below). I think that's a tradeoff that would keep the API consistent, and would allow us to test parity between single-teacher and MOPD.

For example, I tested the scenario below, and the results are identical, except for how we move the teacher from the CPU in the second case.

```python
teacher_model=teachers["a"]    # single-teacher path
teacher_model={"only": teachers["a"]}    # mapping with one entry
```

File: trl/trainer/distillation_trainer.py (hunk 3 above)
Comment (cmpatino) [reply trl/trainer/distillation_trainer.py]: Good catch, thanks!

File: docs/source/paper_index.md (hunk 2 above)
Comment (cmpatino) [reply docs/source/paper_index.md]: That's a good point. I changed the API as I described in [the comment above](https://github.com/huggingface/trl/pull/7215#discussion_r4024553853)

> I agree, so I included the change in `34874dcc`.
> 
> `teacher_model` now takes either a single teacher (unchanged behavior) or a dict[str, str | PreTrainedModel] that routes rows by their teacher_id column. I kept it to dicts and not list/dict, since routing needs the ids.
> 
> I also updated the docstring to emphasize that the type also controls how we handle the teacher. A single teacher keeps the teacher model in the accelerator, while a mapping keeps the teachers on the CPU and moves them one at a time.
> 
> The catch is that we'd also move the teachers to a CPU even if the user specifies a mapping with one model (see the example below). I think that's a tradeoff that would keep the API consistent, and would allow us to test parity between single-teacher and MOPD.
> 
> For example, I tested the scenario below, and the results are identical, except for how we move the teacher from the CPU in the second case.
> 
> ```python
> teacher_model=teachers["a"]    # single-teacher path
> teacher_model={"only": teachers["a"]}    # mapping with one entry
> ```

File: trl/trainer/distillation_trainer.py (hunk 4 above)
Comment (cmpatino) [reply trl/trainer/distillation_trainer.py:2199]: You're right. The paper averages per-trajectory means, while this PR divides the summed divergence by the total number of valid tokens, so longer trajectories weigh more. However, I implemented this version on purpose to keep parity with the single-teacher mode. 

I agree that routing adds a tilt toward the teacher whose domain produces longer completions. However, I'd rather not change the aggregation in this PR to keep it contained to the new feature that keeps parity with the existing behavior.

We can work on a PR that enables choosing the loss type for both multi- and single-teacher paths, as we do in GRPO. The implementation can follow the trajectory count alongside `num_items_in_batch` as you suggest.

File: trl/trainer/distillation_trainer.py [hunk 5]
```diff
@@ -849,13 +955,160 @@ def __init__(
...
+        per_teacher = (self.args.teacher_model_init_kwargs_by_teacher or {}).get(teacher_id, {})
+        init_kwargs = {**common_init_kwargs, **per_teacher}
+        init_kwargs.setdefault("trust_remote_code", self.args.trust_remote_code)
+        # Teachers stay on CPU and are uploaded one at a time, so they are always loaded unmapped.
+        init_kwargs["device_map"] = None
+        commit = None
+        if not isinstance(source, str):
+            teacher, source = source, get_config_model_id(source.config)
+        else:
+            teacher = create_model_from_path(source, **init_kwargs)
```
[the remark this answers is not in the document]
Comment (cmpatino) [reply trl/trainer/distillation_trainer.py:990]: Confirmed, thanks. Fixed in `df69c79f` by mirroring the single-teacher path's `setdefault("revision", ...)`. A test covers both directions through the manifest's resolved commit.

File: trl/trainer/distillation_trainer.py (hunk 4 above)
Comment (diogomsmiranda) [reply trl/trainer/distillation_trainer.py:2199]: lgtm! Once this is merged I can open a draft PR with the option to choose the loss type if you agree.

**Review by kashif — APPROVED** [review]
Looks good to me. The synchronous MOPD implementation is consistent with the existing DistillationTrainer design, preserves the single-teacher path, and has strong unit and distributed coverage.

One small nonblocking improvement that could be good to add: validate that every key in teacher_model_init_kwargs_by_teacher exists in the teacher_model mapping. Currently a typo in a teacher ID is silently ignored, which could cause that teacher to load with the common/default revision or dtype instead of the intended override. This should only require a small set-difference check and a focused test.

File: docs/source/distillation_trainer.md [hunk 6]
```diff
@@ -262,6 +262,92 @@ Tested with:
...
+    model="Qwen/Qwen2.5-0.5B-Instruct",
+    teacher_model={
+        "math": "Qwen/Qwen2.5-Math-1.5B-Instruct",
+        "code": "Qwen/Qwen2.5-Coder-1.5B-Instruct",
+    },
+    train_dataset=dataset,
+    args=DistillationConfig(
+        output_dir="mopd-student",
+        # Route one teacher to a specific revision while the rest load normally.
+        teacher_model_init_kwargs_by_teacher={"code": {"revision": "<revision-name>"}},
```
Comment (qgallouedec) [inline docs/source/distillation_trainer.md:295]: `teacher_model_init_kwargs_by_teacher`? simply `teacher_model_init_kwargs` instead no?

File: docs/source/distillation_trainer.md [hunk 7]
```diff
@@ -262,6 +262,92 @@ Tested with:
...
+    ),
+)
+trainer.train()
+```
+
+### Requirements
+
+- Every teacher must share the student's vocabulary, i.e. the same `vocab_size`. Prompts are rendered once with the student's tokenizer, and every teacher scores those exact token IDs, so a teacher trained on a different tokenizer is training the student against the wrong tokens. Use [GOLD](gold_trainer) for cross-tokenizer distillation.
```
Comment (qgallouedec) [inline docs/source/distillation_trainer.md:303]: just to be precise, I don't think that sharing the same vocab size is enough. It should share the same vocabulary. So imo just
```suggestion
- Every teacher must share the student's vocabulary. Prompts are rendered once with the student's tokenizer, and every teacher scores those exact token IDs, so a teacher trained on a different tokenizer is training the student against the wrong tokens. Use [GOLD](gold_trainer) for cross-tokenizer distillation.
```

File: docs/source/distillation_trainer.md [hunk 8]
```diff
@@ -262,6 +262,92 @@ Tested with:
...
+- Teachers must be unquantized teachers loaded without device dispatch or offload hooks. Teachers move between CPU and the accelerator as complete modules, and the loss reads their output heads as dense matrices; quantized and dispatched teacher backends have not been integrated or validated. A `quantization_config` and an active `device_map` are rejected at initialization, per teacher after the per-teacher overrides are merged, and every loaded teacher is checked for quantization and device dispatch — including one loaded from a checkpoint that carries its quantization in its own config. `device_map=None` is allowed.
```
Comment (qgallouedec) [inline docs/source/distillation_trainer.md:304]:
```suggestion
- Teachers must be unquantized loaded without device dispatch or offload hooks. Teachers move between CPU and the accelerator as complete modules, and the loss reads their output heads as dense matrices; quantized and dispatched teacher backends have not been integrated or validated. A `quantization_config` and an active `device_map` are rejected at initialization, per teacher after the per-teacher overrides are merged, and every loaded teacher is checked for quantization and device dispatch — including one loaded from a checkpoint that carries its quantization in its own config. `device_map=None` is allowed.
```

File: tests/distributed/test_distillation_trainer_multi_teacher.py
Comment (qgallouedec) [inline tests/distributed/test_distillation_trainer_multi_teacher.py:1]: I reviewed this test file and the other, and I think it's not how we want to test that. Mostly because we test on GPU (here it's trying to test on CPU), and it contains flaky patterns like the `SimpleNamespace`, and private method testing that we try to avoid as much as possible.
So let's remove these two files. And instead either:
- add a test case in `test_distributed.py` following the existing convention
- or, if it's to hard, because you need to change the script or I don't know, let's just add this in a follow-up PR.

File: trl/trainer/distillation_trainer.py [hunk 9]
```diff
@@ -396,7 +406,7 @@ class DistillationTrainer(_BaseTrainer):
     def __init__(
         self,
         model: "str | PreTrainedModel | PeftModel",
-        teacher_model: str | PreTrainedModel | None = None,
+        teacher_model: "str | PreTrainedModel | dict[str, str | PreTrainedModel] | None" = None,
```
Comment (qgallouedec) [inline trl/trainer/distillation_trainer.py:409]: why not

```suggestion
        teacher_model: str | PreTrainedModel | dict[str, str | PreTrainedModel] | None = None,
```
?

**Review by qgallouedec — COMMENTED** [review]
Initial review, I'll need more time
